File: 55. Jump Game.java
enum Index {
    GOOD, BAD, UNKNOWN
}

class Solution {
    public boolean canJump(int[] nums) {
        int lastPos = nums.length - 1;
        for (int i = nums.length - 1; i >= 0; i--) {
            if (nums[i] + i >= lastPos) {
                lastPos = i;
            }
        }
        return lastPos == 0;
    }
    
//     public boolean canJump(int[] nums) {
//         Index[] memo = new Index[nums.length];
//         for (int i = 0; i < memo.length; i++) {
//             memo[i] = Index.UNKNOWN;
//         }
//         memo[memo.length-1] = Index.GOOD;
        
//         for (int i = nums.length - 2; i >= 0; i--) {
//             int furthestJump = Math.min(i + nums[i], nums.length - 1);
//             for (int j = i + 1; j <= furthestJump; j++) {
//                 if (memo[j] == Index.GOOD) {
//                     memo[i] = Index.GOOD;
//                     break;
//                 }
//             }
//         }
        
//         return memo[0] == Index.GOOD;
//     }
    
//     Index[] memo;
    
//     public boolean canJump(int[] nums) {
//         memo = new Index[nums.length];
//         for (int i = 0; i < memo.length; i++) {
//             memo[i] = Index.UNKNOWN;
//         }
//         memo[memo.length-1] = Index.GOOD;
//         return canJumpFromPosition(0,nums);
//     }
    
//     public boolean canJumpFromPosition(int position, int[] nums) {
//         if (memo[position] != Index.UNKNOWN) {
//             return memo[position] == Index.GOOD ? true : false;
//         }
        
//         int furthestJump = Math.min(nums.length - 1, position + nums[position]);
        
        
//         //if (position == nums.length - 1) return true;
        
//         for(int nextPosition = position + 1; nextPosition <= furthestJump; nextPosition++) {
//             if (canJumpFromPosition(nextPosition, nums)) {
//                 memo[position] = Index.GOOD;
//                 return true; // break out of the loop 
//             }
//         }
//         memo[position] = Index.BAD;
//         return false;
//     }
}


// a really good article: 
// https://leetcode.com/problems/jump-game/solution/

File: amazon/python/55. Jump Game.py
class Solution(object):
    def canJump(self, nums):
        lastPos = len(nums) - 1
        for i in range(len(nums) - 2, -1, -1):
            far_pos = min(i + nums[i], len(nums) -1)
            if far_pos >= lastPos:
                lastPos = i
        return lastPos == 0

    # Time: O(n)
    # space: O(1)
    
#     def canJumpFromPosition(self, position, nums, memo):
#         if memo[position] != 'unknown':
#             if memo[position] == 'good':
#                 return True
#             else:
#                 return False
#         far_position = min(position + nums[position], len(nums) - 1)
#         for nextPos in range(position+1, far_position+1):
#             if self.canJumpFromPosition(nextPos, nums, memo):
#                 memo[position] = 'good'
#                 return True
#         memo[position] = 'bad'
#         return False
    
#     def canJump(self, nums):
#         memo = ['unknown' for _ in range(len(nums))]
#         memo[len(memo) - 1] = 'good'
#         return self.canJumpFromPosition(0, nums, memo)
    
    
    # Time: O(n)
    # Space: O(1)
#     def canJump(self, nums):
#         """
#         :type nums: List[int]
#         :rtype: bool
#         """
#         return self.canJumpFromPosition(0, nums)
    
    
#     def canJumpFromPosition(self, position, nums):
#         if position == len(nums) - 1:
#             return True
#         far_position = min(position + nums[position], len(nums) - 1)
#         for next_position in range(far_position, position, -1):
#             if self.canJumpFromPosition(next_position, nums):
#                 return True
#         return False
    
    ## Time: O(2^n)
    ## Space: O(n)
    
